packages/@aws-cdk/aws-cognito/README.md

```diff
@@ -753,3 +766,374 @@ Existing domains can be imported into CDK apps using `UserPoolDomain.fromDomainN
+
+Like the supported external providers, though, only one custom provider can be directly associated with the identity  
+pool.
+
+```ts
+new cognito.IdentityPool(this, 'myidentitypool', {
+  identityPoolName: 'myidentitypool',
+  grantUserActions: ['execute-api'],
+  authenticationProviders: {
+    google: '12345678012.apps.googleusercontent.com',
+    openIdConnectProvider: openIdConnectProvider,
+    customProvider: 'my-custom-provider.example.com',
+  }
+});
+```
+
+### Role Mapping
+
+In addition to setting default roles for authenticated and unauthenticated users, identity pools can also be used to  
+define rules to choose the role for each user based on claims in the user's ID token by using Role Mapping. When using  
+role mapping, it's important to be aware of some of the permissions the role will need. An in depth  
+review of roles and role mapping can be found [here](https://docs.aws.amazon.com/cognito/latest/developerguide/role-based-access-control.html).
+
+Using a [token-based approach](https://docs.aws.amazon.com/cognito/latest/developerguide/role-based-access-control.html#using-tokens-to-assign-roles-to-users) to role mapping will allow mapped roles to be passed through the `cognito:roles` or  
+`cognito:preferred_role` claims from the identity provider:
+
+```ts
+new cognito.IdentityPool(this, 'myidentitypool', {
+  identityPoolName: 'myidentitypool',
+  roleMappings: [{
+    providerUrl: SupportedLoginProviderType.AMAZON,
+    useToken: true,
+  }],
+});
+```
+
+Using a rule-based approach to role mapping allows roles to be assigned based on custom claims passed from the identity  provider:
+
+```ts
+new cognito.IdentityPool(this, 'myidentitypool', {
+  identityPoolName: 'myidentitypool',
+  // Assign specific roles to users based on whether or not the custom admin claim is passed from the identity provider
+  roleMappings: [{
+    providerUrl: SupportedLoginProviderType.AMAZON,
+    rules: [
+      {
+        claim: 'custom:admin',
+        claimValue: 'admin',
+        mappedRole: adminRole,
+      },
+      {
+        claim: 'custom:admin',
+        claimValue: 'admin',
+        matchType: RoleMappingMatchType.NOTEQUAL,
+        mappedRole: nonAdminRole,
+      }
+    ],
+  }],
+});
+```
+
+Role mappings can also be added after instantiation with the Identity Pool's `addRoleMappings` method:
+
+```ts
+identityPool.addRoleMappings(myAddedRoleMapping1, myAddedRoleMapping2, myAddedRoleMapping3);
+```
+
+### Identity Pool Role Attachments
+
+The `IdentityPoolRoleAttachment` Construct represents all roles attached to an identity pool. This includes default  
+roles and role mappings. The `IdentityPool` Construct uses `IdentityPoolRoleAttachments` under the hood to add default  
+roles and role mappings, so it's usually not necessary to use `IdentityPoolRoleAttachments` directly. The one use case  
+in which it may come in handy, however, is in order to replace the default roles of an existing identity pool:
```

Comment:
There are a few rare use cases, but I believe the changes suggested directly below would render them unnecessary, so I'll drop this. Should I move the `IdentityPoolRoleAttachment` to the `private` directory since it won't be used directly?